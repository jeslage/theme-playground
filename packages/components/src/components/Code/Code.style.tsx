import styled from 'styled-components';

const StyledCode = styled.div`
  margin: 1.5rem 1rem;

  .code__wrapper {
    position: relative;
  }

  code {
    display: block;
    margin: 0;
    background: none;
    color: ${({ theme }) => theme.text};
    border: 1px solid ${({ theme }) => theme.border};
    border-radius: 5px;
  }

  pre {
    margin: 0;
    padding: 1rem;
    font-size: 14px;
  }

  button {
    position: absolute;
    top: 0;
    right: 0;
    margin: 0;
    outline: none;
    border: none;
    padding: 5px 10px;
    border-radius: 0px 5px 0px 5px;
    font-weight: bold;
    font-size: 12px;
    cursor: pointer;
    background: none;
    color: ${({ theme }) => theme.text};
    border: 1px solid ${({ theme }) => theme.border};

    &:active {
      color: ${({ theme }) => theme.text};
    }

    &:disabled {
      opacity: 1;
    }
  }
`;

export default StyledCode;
