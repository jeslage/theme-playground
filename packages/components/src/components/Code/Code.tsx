import * as React from 'react';

import StyledCode from './Code.style';

interface CodeProps {
  value: object;
}

const Code: React.FC<CodeProps> = ({ value }) => {
  const [copied, setCopied] = React.useState(false);

  const copyToClipboard = () => {
    const element = document.createElement('textarea');
    element.value = JSON.stringify(value, null, 2);
    element.setAttribute('readonly', '');
    element.style.position = 'absolute';
    element.style.left = '-9999px';
    document.body.appendChild(element);
    element.select();
    if (document.execCommand) {
      document.execCommand('copy');
    }
    document.body.removeChild(element);
    setCopied(true);
  };

  React.useEffect(() => {
    const timeout = setTimeout(() => {
      setCopied(false);
    }, 1500);
    return () => {
      clearTimeout(timeout);
    };
  }, [copied]);

  return (
    <StyledCode>
      <div className="code__wrapper">
        <code>
          <pre>{JSON.stringify(value, null, 2)}</pre>
        </code>
        <button type="button" onClick={copyToClipboard} disabled={copied}>
          {copied ? 'Copied' : 'Copy'}
        </button>
      </div>
    </StyledCode>
  );
};

export default Code;
