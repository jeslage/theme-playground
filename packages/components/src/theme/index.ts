export const darkTheme = {
  base: 'dark',
  backgroundDark: '#2f2f2f',
  backgroundLight: '#333',
  border: '#999999',
  text: '#FFFFFF'
};

export const lightTheme = {
  base: 'light',
  backgroundDark: '#f6f9fc',
  backgroundLight: '#fff',
  border: '#666666',
  text: '#333333'
};

export const getUiTheme = theme => {
  if (theme === undefined) return darkTheme;

  if (typeof theme === 'string') {
    if (theme === 'dark') {
      return darkTheme;
    } else if (theme === 'light') {
      return lightTheme;
    }

    throw Error(
      `Theme "${theme}" is not available. Please use "light" or "dark", or provide a custom theme object.`
    );
  }

  if (theme.base && theme.base === 'light') {
    return {
      ...lightTheme,
      ...theme
    };
  }

  return {
    ...darkTheme,
    ...theme
  };
};
