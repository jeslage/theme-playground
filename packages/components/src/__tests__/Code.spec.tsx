import 'jest';
import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { ThemeProvider } from 'styled-components';

import { darkTheme } from '../theme';
import Code from '../components/Code/Code';
import StyledCode from '../components/Code/Code.style';

describe('Code', () => {
  it('Code should render correctly', () => {
    const component = shallow(<Code value={{ theme: { color: 'red' } }} />);

    expect(component).toMatchSnapshot();
  });

  it('StyledCode should render correctly', () => {
    const component = shallow(<StyledCode theme={{ colors: darkTheme }} />);

    expect(component).toMatchSnapshot();
  });

  it('Code should be copied', () => {
    const component = mount(
      <ThemeProvider theme={{ colors: darkTheme }}>
        <Code value={{ theme: { color: 'red' } }} />
      </ThemeProvider>
    );

    component.find('button').simulate('click');

    expect(component).toMatchSnapshot();
  });
});
