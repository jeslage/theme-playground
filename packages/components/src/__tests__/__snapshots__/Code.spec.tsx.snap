// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code Code should be copied 1`] = `
<ThemeProvider
  theme={
    Object {
      "colors": Object {
        "backgroundDark": "#2f2f2f",
        "backgroundLight": "#333",
        "base": "dark",
        "border": "#999999",
        "text": "#FFFFFF",
      },
    }
  }
>
  <Code
    value={
      Object {
        "theme": Object {
          "color": "red",
        },
      }
    }
  >
    <styled.div>
      <div
        className="sc-AxjAm brjcLn"
      >
        <div
          className="code__wrapper"
        >
          <code>
            <pre>
              {
  "theme": {
    "color": "red"
  }
}
            </pre>
          </code>
          <button
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Copied
          </button>
        </div>
      </div>
    </styled.div>
  </Code>
</ThemeProvider>
`;

exports[`Code Code should render correctly 1`] = `
<styled.div>
  <div
    className="code__wrapper"
  >
    <code>
      <pre>
        {
  "theme": {
    "color": "red"
  }
}
      </pre>
    </code>
    <button
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Copy
    </button>
  </div>
</styled.div>
`;

exports[`Code StyledCode should render correctly 1`] = `
<div
  className="sc-AxjAm brjcLn"
/>
`;
